Change 'rabbitmqctl status' to a wget | grep to save CPU

wanderboessenkool opened the pull request:
- This reduces CPU usage from 250 millis on idle to 25 millis on idle
- Default rabbitmq user needs administrator privileges

##### SUMMARY
In the default configuration for the awx-rabbit container, the healthchekcs (livenessProbe and readinessProbe) use the comand 'rabbitmqctl status'. This causes high CPU usage, even on idle (250 millis). By changing the healthchecks to an authenticated request to http://localhost:15672/api/healthchecks/node, and grepping for the desired result in the output, this can be reduced to less than 25 millis, a factor ten saving.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
 - Installer

##### AWX VERSION
```
awx: 7.0.0
```


##### ADDITIONAL INFORMATION
There are other projects suffering the same from the CPU hungry behavior of `rabbitmqctl status`, see:
- [GCP Issue](https://github.com/GoogleCloudPlatform/click-to-deploy/issues/522)
- [Helm Issue](https://github.com/helm/charts/issues/3855)

wanderboessenkool committed:
Change 'rabbitmqctl status' to a wget | grep

- This reduces CPU usage from 250 millis on idle to 25 millis on idle
- Default rabbitmq user needs administrator privileges

wenottingham commented:
What ensures wget and ash are in the base container image?

wanderboessenkool committed:
Properly escape quotes

wanderboessenkool commented:
> What ensures wget and ash are in the base container image?

wget is provided by busybox, as is ash. They are both present in the official awx-rabbit image, which pulls from rabbitmq:${RABBITMQ_VERSION}-management-alpine, which pulls from rabbitmq:${RABBITMQ_VERSION}-alpine, which pulls from alpine:3.10, which has both installed by default.

When you put in your own {{ kubernetes_rabbitmq_image }} it would depend. I can switch the shell to /bin/sh or /bin/bash, since they are present as well in the same image and probably present in more base images than ash.

If you want to get rid of wget it will have to be replaced by curl, or a custom python script that gets added to the image.

ryanpetrello reviewed (commented):
@matburt you have any thoughts on this?

matburt commented:
This is pretty compelling! I'd be willing to merge this if you can show you verification methodology, how you sampled usage, and provide me a way to verify it.

wanderboessenkool commented:
> This is pretty compelling! I'd be willing to merge this if you can show you verification methodology, how you sampled usage, and provide me a way to verify it.

Without the change, with `rabbitmqctl status`:                                            
                                                                                          
`oc adm top pod --containers`                                                             
```                                                                                       
POD                  NAME            CPU(cores)   MEMORY(bytes)                      
awx-0                awx-memcached   0m           7Mi                        
awx-0                awx-web         0m           552Mi              
awx-0                awx-celery      5m           458Mi                    
awx-0                awx-rabbit      206m         102M
postgresql-2-8w8qd   postgresql      4m           87Mi
```                                                                                       
                                                                                          
With the change to http based healthchecks, grepping for `{"status":"ok"}`:               
                                                                                          `oc adm top pod --containers`                                                             
```                                                                                       
POD                  NAME            CPU(cores)   MEMORY(bytes)
awx-0                awx-memcached   0m           7Mi                        
awx-0                awx-web         0m           552Mi                                   
awx-0                awx-celery      5m           702Mi                                   
awx-0                awx-rabbit      16m          118Mi
postgresql-2-8w8qd   postgresql      5m           91Mi
```                                                                                                                                                                                 
In prometheus:                                                                            `rate(container_cpu_usage_seconds_total{namespace="awx", container_name="awx-rabbit"}[1m])`                                                                                         
(substituting `namespace="awx"` for whichever namespace awx is running in.)

shanemcd reviewed (changes requested):
I am all for this change but we can't use `/bin/ash` because our downstream images are RHEL-based. Please switch to `/bin/sh` or `/bin/bash` then I'll +1.

wanderboessenkool committed:
Change /bin/ash to /bin/sh as requested by @shanecmd

wanderboessenkool committed:
Change healthcheck from wget and grep to python with httplib

wanderboessenkool commented:
@shanecmd. I've changed `/bin/ash` to `/bin/sh`.

Looking at the downstream image they do not seem to have wget installed, but they do have curl. That doesn't help in trying to unify the two. I've added another change that moves the healthcheck into a small inline python program.

Going forward that healthcheck could be moved into a separate script file, but it would have to be added to both the awx and the downstream images.

shanemcd reviewed (approved):
This is really impressive stuff. Thank you for the solid work and quick turnaround time.

shanemcd commented:
I just noticed that we're templating out the username / password in the deployment object itself. Can we move this to read from an env var?

ryanpetrello commented on installer/roles/kubernetes/templates/deployment.yml.j2:
```diff
@@ -269,12 +269,44 @@ spec:
+              - /usr/bin/python
+              - -c
+              - |
+                import httplib
```
@wanderboessenkool this assumes that `/usr/bin/python` is python2, which may not be a safe assumption to make downstream cc @shanemcd

ryanpetrello replied:
```python
try:
    from http.client import HTTPConnection
except ImportError:
    from httplib import HTTPConnection
```

wanderboessenkool replied:
I've modified the import according to your example. Thanks for spotting it.

wanderboessenkool committed:
Move installtime hardcoded rabbitmq credentials to environment variables for healthcheck

wanderboessenkool commented:
> I just noticed that we're templating out the username / password in the deployment object itself. Can we move this to read from an env var?

Yes we can :-)

ryanpetrello reviewed (changes requested):
see: https://github.com/ansible/awx/pull/5009/files#r336229304

shanemcd commented:
Apologies for the noisy feedback, juggling a lot atm and just returned from vacation today. Another thought: we could do something like [this](https://github.com/ansible/awx/blob/devel/installer/roles/kubernetes/templates/deployment.yml.j2#L40) to pull this into a ConfigMap / mounted file which would be cleaner than inlining the whole script.

wanderboessenkool committed:
Make HTTPConnection import python 2,3 agnostic

ryanpetrello commented on installer/roles/kubernetes/templates/deployment.yml.j2:
```diff
@@ -269,12 +269,56 @@ spec:
+              - -c
+              - |
+                try:
+                  from http.client import HTTPConnection
```
👍

wanderboessenkool commented:
> Apologies for the noisy feedback, juggling a lot atm and just returned from vacation today. Another thought: we could do something like [this](https://github.com/ansible/awx/blob/devel/installer/roles/kubernetes/templates/deployment.yml.j2#L40) to pull this into a ConfigMap / mounted file which would be cleaner than inlining the whole script.

@shanemcd  That would definitely be nicer on the yaml for the healthchecks, but it might make it harder for people to debug. If wanted I can make those changes tomorrow.

wanderboessenkool committed:
Move python healthcheck script from probes to configMap

shanemcd commented:
Thanks for this!
